### PR TITLE
Add ClientTimeout to LeaderElection configuration

### PR DESCRIPTION
**Pull request by jprzychodzen**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This change adds additional parameter to leader election configuration and a new path to create resource lock in a backward compatible. Currently, timeout for kube-apiserver client is set in a library's client dependent way, allowing setting unreasonable values like RenewDeadline.

Too high value for ClientTimeout resulted in a fragile leader election, failing with only a single timeout.

This PR fixes unreasonable values for current usage and provides a unified and parameterized way to set client timeout for leader election's kube-apiserver clients.

**Does this PR introduce a user-facing change?**:
```release-note
New parameter for leader election - ClientTimeout. Action required for client-go client if they want to use this parameter, switch to new configuration method needed.
```

### CODE REVIEW

File: staging/src/k8s.io/component-base/config/types.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ type ClientConnectionConfiguration struct {
 // LeaderElectionConfiguration defines the configuration of leader election
 // clients for components that can run with leader election enabled.
 type LeaderElectionConfiguration struct {
+	// clientTimeout indicates the timeout set for client used to communicate
+	// with kube-apiserver.
+	ClientTimeout metav1.Duration
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/config/types.go:44]: I think that overall this parameter makes sense to me.
However, I'm not convinced that it should really be defaulted to 2s (in defaulting).

In my personal opinion what we should do is:
- add the field here
- default it to somewhat what we were previously doing (so either to RenewDeadline, or maybe 10s which is its default value)
- start respecting it

And in the second PR, switch it to 2s (or whatever else value) in kcm, ccm and scheduler.

But given it's an API change, I would like to hear @liggitt `s thoughts on that too.

**Comment by wojtek-t** [comment]
/assign @liggitt 

@liggitt - can you please share your thoughts on https://github.com/kubernetes/kubernetes/pull/95319#discussion_r503244110 (and in particular on the API part of it of couse)

**Comment by liggitt** [comment]
> Too high value for ClientTimeout resulted in a fragile leader election, failing with only a single timeout.

Is that accurate? We're setting the request-specific timeout in LeaderElector#renew here:

https://github.com/kubernetes/kubernetes/blob/a9bfd8816d860b9585b65471b0ef911ccb6054d3/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L262-L267

If the clientset's timeout is longer, the request-specific timeout should still short-circuit the request early and safely.

I don't see a reason we need configuration and replumbing for this... I'd expect to be able to use the existing context timeout functionality the clientset provides.

**Comment by wojtek-t** [comment]
> If the clientset's timeout is longer, the request-specific timeout should still short-circuit the request early and safely.

@liggitt That is true. But the problem is that single "renew" operation actually has multiple API calls underneath.
And we've observed case that one of them were stuck until timeout because of ?network? issues. And that was meaning we were losing leader lock.

We need a way to set the timeout on the inidvidual api call level lower than that to ensure that we have a chance to retry without loosing the lock.

**Comment by liggitt** [comment]
> We need a way to set the timeout on the inidvidual api call level lower than that to ensure that we have a chance to retry without loosing the lock.

that still seems like can be done internally without additional configuration, though it's not clear to me that shortening timeouts for sub-requests in order to accomodate retries will not trigger new failure and leader-loss modes.

For example, say the renew deadline is 10 seconds, there are 4 requests to make (multi-object lock: get primary, get secondary, create/update primary, create/update secondary), and each request takes 2 seconds for the server to respond to.

With the current code the requests would all complete in 8 seconds.

If we artificially shortened the timeout to give sufficient time for retries, that would give each request a shortened 1.25 second timeout (4 requests with retries = 8 requests = 1.25 seconds per request to complete in 10 seconds). If the request takes 2 seconds for the server to satisfy, this means all our requests will time out, when they would not have before.

**Comment by wojtek-t** [comment]
> that still seems like can be done internally without additional configuration, though it's not clear to me that shortening timeouts for sub-requests in order to accomodate retries will not trigger new failure and leader-loss modes.

it can be done internally, but those "other failure modes" are exactly why I think this should be exposed (with defaulted to what we have now).
Currently, with timeout set to 10s, one "stuck" API calls results in loss of leader.
Given that generally requests finish very fast (in small hundreds of millisecond), setting it to e.g. 5s means that it will not significantly reduce the amount of cases that currently could work. OTOH, it would fix the cases where requests is stuck until timeout, which we've seen.

In my personal we should set it the way that guarantees that 8 requests (or N requests in general) would finish on time. But we should set it for something, that a single timeout and reasonably fast other requests could work. I think 50% of renewDeadline is what I would personally suggest. But I'm not sure I would like this to be hidden and so that users can actually set that differently if they want.

But if you think that 50% hardcoded in the library below is reasonable enough, I would be fine with that too.

**Comment by jprzychodzen** [comment]
Comment about number of requests - for multi-object lock case there is 5, not 4 API calls. Update makes an additional GET for a secondary resource

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/multilock.go#L78

I'm not exactly sure how one will end up using client-go library, what would be network environment and I do not want to restrict that. Correlating client timeout with other parameters like renew deadline or retry period in a hardcoded way will break some use cases. However, this parameter must be changed or at least allowed to be tweakable.

Setting it as a 50% of renew period with low retry period will kind of break what was expected - one can assume that there will be RenewDeadline/RetryPeriod chances (defaulting to 5 for 10s and 2s), but actually for client timeout equal 50% of retry period there will be exactly 2 chances to renew leader election lock. Setting RetryPeriod to even lower values will only speed up a process of getting to two sequential failures for lock renewal.

If some parameter should be hardcoded internally, then it should also consider RetryPeriod - something like MIN(RetryPeriod, RenewDeadline/2) is a better mach for a client timeout. However, we are slowly starting to sink in covering different edge cases, so I think it's best to allow affected user to decide how to handle different failure modes based on environment and configuration.

**Comment by wojtek-t** [comment]
> Setting it as a 50% of renew period with low retry period will kind of break what was expected - one can assume that there will be RenewDeadline/RetryPeriod chances (defaulting to 5 for 10s and 2s), but actually for client timeout equal 50% of retry period there will be exactly 2 chances to renew leader election lock.

I don't buy this argument, because you don't have those guarantees now too.
Even the API call latency SLO we have says implicitly that requests may take more than 1s.
So you may have the same problem as of now, even if everything works perfectly fine.

**Comment by jprzychodzen** [comment]
> Even the API call latency SLO we have says implicitly that requests may take more than 1s.
I feel that we are already moving in a space of overloaded kube-apiserver, where we do not really count for that "1s" resposne.

> So you may have the same problem as of now, even if everything works perfectly fine.
I'm not exactly sure what you mean by that, could you elaborate?

What I'm saying is that RenewDeadline has different purpose than RetryPeriod. RetryPeriod is more related to what we should use.

RenewDeadline is correlated to guarantees of a software, how long it can survive without a leader and how it interacts with other components - the is externally visible behavior.

RetryPeriod is related how we achieve targeted RenewDeadline. It looks like an internal parameter that we could use to tweak how thing works internally. It just make more sense for me to tie client timeout with internal retry period instead of externally affecting renew deadline.

So if we want to hardcode this value somehow, I guess that it's reasonable to set it to RetryPeriod (or some const multipiler of it, like 2*RetryPeriod). This way software can achieve it's targeted guaranties for time between leader elections, and there is a possibility to tweak client timeout to match environment/configuration.

**Comment by liggitt** [comment]
> RetryPeriod is related how we achieve targeted RenewDeadline. It looks like an internal parameter that we could use to tweak how thing works internally. It just make more sense for me to tie client timeout with internal retry period instead of externally affecting renew deadline.

Right now RetryPeriod is the interval between completion of one attempt and beginning the next (the retry is sliding). What you're describing sounds more like making it a non-sliding interval, is that correct?

**Comment by jprzychodzen** [comment]
I didn't want to suggest it, but yes was thinking about 'tryAcquireOrRenew' as a quick operation. In a healthy state it's rather quick operation and we repeat tryAcquireOrRenew every 2 seconds .

Changing it to non-sliding wasn't something what I'm suggesting - either we end up with multiple 'tryAcquireOrRenew' happening in parallel, or we need to cancel previous context after RetryPeriod seconds, and I do not think that any of those ideas sound reasonable.

We could try non-sliding interval and cancelling context only if we have just finished PUT API call (we invalidated results of other API calls). It would complicate current approach a lot, because we would need to keep goroutine-state of each lock for each 'tryAcquireOrRenew' context and manage state between different steps.

Adding '--client-timeout' parameter sounds like a less complicated, tunable and working solution to fragile leader election.

**Comment by jprzychodzen** [comment]
After thinking, it seems like a superior solutions - it gives a maximal available timeout equal to RenewDeadline, and is backward compatible.

It will be a little tricky to code, so please say what do you think about this approach.

@wojtek-t @liggitt

**Comment by wojtek-t** [comment]
FTR - we super briefly discussed with Jordan, and the outcome was roughly this:

"I don't think an additional "individual request timeout" knob for clients is going to improve things in general
I pretty much agree with something like max(time.Second, RenewDeadline/2) to keep a single hung request from forcing a leader loss""

So we will proceed with the approach described above. https://github.com/kubernetes/kubernetes/pull/95646 is the first step towards this.

**Comment by jprzychodzen** [comment]
/close
PR not needed anymore. Part of it is already merged, remaining code will be added in component specific PRs.

**Pull request closed without merging**